impl/src/main/java/org/ehcache/internal/events/ScopedStoreEventDispatcher.java

```diff
@@ -0,0 +1,313 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.internal.events;
+
+import org.ehcache.event.EventType;
+import org.ehcache.events.StoreEventDispatcher;
+import org.ehcache.events.StoreEventSink;
+import org.ehcache.spi.cache.events.StoreEvent;
+import org.ehcache.spi.cache.events.StoreEventFilter;
+import org.ehcache.spi.cache.events.StoreEventListener;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.ehcache.internal.events.StoreEvents.createEvent;
+import static org.ehcache.internal.events.StoreEvents.evictEvent;
+import static org.ehcache.internal.events.StoreEvents.expireEvent;
+import static org.ehcache.internal.events.StoreEvents.removeEvent;
+import static org.ehcache.internal.events.StoreEvents.updateEvent;
+
+/**
+ * ScopedStoreEventDispatcher
+ */
+public class ScopedStoreEventDispatcher<K, V> implements StoreEventDispatcher<K, V> {
+
+  private static final StoreEventSink NO_OP_EVENT_SINK = new StoreEventSink() {
+    @Override
+    public void removed(Object key, Object value) {
+      // Do nothing
+    }
+
+    @Override
+    public void updated(Object key, Object oldValue, Object newValue) {
+      // Do nothing
+    }
+
+    @Override
+    public void expired(Object key, Object value) {
+      // Do nothing
+    }
+
+    @Override
+    public void created(Object key, Object value) {
+      // Do nothing
+    }
+
+    @Override
+    public void evicted(Object key, Object value) {
+      // Do nothing
+    }
+  };
+
+  private final Set<StoreEventFilter<K, V>> filters = new CopyOnWriteArraySet<StoreEventFilter<K, V>>();
+  private final Set<StoreEventListener<K, V>> listeners = new CopyOnWriteArraySet<StoreEventListener<K, V>>();
+  private final BlockingQueue<FireableStoreEventWrapper<K, V>>[] orderedQueues = new LinkedBlockingQueue[4];
+  private volatile boolean ordered = false;
+
+  public ScopedStoreEventDispatcher() {
+    for (int i = 0; i < orderedQueues.length; i++) {
+      orderedQueues[i] = new LinkedBlockingQueue<FireableStoreEventWrapper<K, V>>(10000);
+    }
+  }
+
+  @Override
+  public StoreEventSink<K, V> eventSink() {
+    if (listeners.isEmpty()) {
+      return NO_OP_EVENT_SINK;
+    } else {
+      return new InvocationScopedEventSink<K, V>(filters, ordered, orderedQueues);
+    }
+  }
+
+  @Override
+  public void releaseEventSink(StoreEventSink<K, V> eventSink) {
+    if (eventSink != NO_OP_EVENT_SINK) {
```

Comment:
Isn't this spilling the guts of InvocationScopedEventSink on to the table?  The if check here is basically a type check right?